### PR TITLE
Allow pulumi stack export to decrypt secrets

### PR DESCRIPTION
**Pull request by stack72**
Fixes: #2918

This allows us to run the command `pulumi stack export --show-secrets`
it will also introduce the changes that allows the import to handle
when plain text is included in the import file

### CODE REVIEW

**Comment by stack72** [comment]
@lukehoban / @pgavlin would love eyes on this when you get some time please

File: pkg/backend/httpstate/backend.go [hunk 1]
```diff
@@ -1177,18 +1178,47 @@ func (b *cloudBackend) ExportDeploymentForVersion(
...
 	}
 
-	return b.exportDeployment(ctx, stack.Ref(), &versionNumber)
+	return b.exportDeployment(ctx, stack, &versionNumber, showSecrets)
 }
 
 // exportDeployment exports the checkpoint file for a stack, optionally getting a previous version.
 func (b *cloudBackend) exportDeployment(
-	ctx context.Context, stackRef backend.StackReference, version *int) (*apitype.UntypedDeployment, error) {
+	ctx context.Context, stk backend.Stack, version *int, showSecrets bool) (*apitype.UntypedDeployment, error) {
```
Comment (lukehoban) [inline pkg/backend/httpstate/backend.go]: This is just in the cloud backend.  Do we have the ability to do this higher in the stack so that we don't have to implement separately for both backends?  It feels like at least in theory this shouldn't have anything to do with the backend.

File: pkg/resource/stack/deployment.go [hunk 2]
```diff
@@ -482,19 +501,32 @@ func DeserializePropertyValue(v interface{}, dec config.Decrypter) (resource.Pro
...
-					plaintext, err := dec.DecryptValue(ciphertext)
-					if err != nil {
-						return resource.PropertyValue{}, errors.Wrap(err, "decrypting secret value")
+
+					if plainOk {
+						encryptedText, err := enc.EncryptValue(plaintext)
+						if err != nil {
+							return resource.PropertyValue{}, errors.Wrap(err, "encrypting secret value")
+						}
+						ciphertext = encryptedText
```
Comment (lukehoban) [inline pkg/resource/stack/deployment.go]: Why do we actually have to do this encryption and compute this ciphertext?

It seems the only place this is used is to populate the encryption cache - but that shouldn't be needed if we don't even have to call Encrypt in the first place to compute this.

If we could avoid this, it also avoids plumbing an encyrpter through all of this code.

File: pkg/backend/httpstate/backend.go [hunk 3]
```diff
@@ -1177,18 +1178,47 @@ func (b *cloudBackend) ExportDeploymentForVersion(
...
+
+		data, err := json.Marshal(sdep)
+		if err != nil {
+			return nil, err
+		}
+
+		return &apitype.UntypedDeployment{
+			Version:    3,
+			Deployment: json.RawMessage(data),
+		}, nil
```
Comment (lukehoban) [inline pkg/backend/httpstate/backend.go]: This doesn't feel quite right - it feels like this codepath should be more similar to the non-showSecrets path - and that the differences should be more clear in another layer - ideally a higher layer so this can be reused across backends, but if not then maybe even a lower layer?

File: pkg/backend/httpstate/state.go [hunk 4]
```diff
@@ -261,8 +261,8 @@ func (b *cloudBackend) newUpdate(ctx context.Context, stackRef backend.StackRefe
 	}, nil
 }
 
-func (b *cloudBackend) getSnapshot(ctx context.Context, stackRef backend.StackReference) (*deploy.Snapshot, error) {
-	untypedDeployment, err := b.exportDeployment(ctx, stackRef, nil /* get latest */)
+func (b *cloudBackend) getSnapshot(ctx context.Context, stk backend.Stack) (*deploy.Snapshot, error) {
+	untypedDeployment, err := b.exportDeployment(ctx, stk, nil /* get latest */, false)
```
Comment (lukehoban) [inline pkg/backend/httpstate/state.go]: Was there a reason for this move from `StackReference` to `Stack`?  I thought we moved the other direction on this API recently? 

Also - why `stk`?

File: cmd/stack_import.go [hunk 5]
```diff
@@ -130,7 +130,7 @@ func newStackImportCmd() *cobra.Command {
 
 				snapshot.PendingOperations = nil
 			}
-			sdp, err := stack.SerializeDeployment(snapshot, snapshot.SecretsManager)
+			sdp, err := stack.SerializeDeployment(snapshot, snapshot.SecretsManager, false /** showSecrets**/)
```
Comment (pgavlin) [inline cmd/stack_import.go]:
```suggestion
			sdp, err := stack.SerializeDeployment(snapshot, snapshot.SecretsManager, false /* showSecrets */)
```

File: cmd/stack_output.go [hunk 6]
```diff
@@ -92,7 +92,8 @@ func newStackOutputCmd() *cobra.Command {
 	cmd.PersistentFlags().BoolVarP(
 		&jsonOut, "json", "j", false, "Emit output as JSON")
 	cmd.PersistentFlags().StringVarP(
-		&stackName, "stack", "s", "", "The name of the stack to operate on. Defaults to the current stack")
+		&stackName, "stack", "s", "",
```
Comment (pgavlin) [inline cmd/stack_output.go]: Why was this changed?

File: cmd/state.go [hunk 7]
```diff
@@ -167,7 +167,7 @@ func runTotalStateEdit(
 		contract.AssertNoErrorf(snap.VerifyIntegrity(), "state edit produced an invalid snapshot")
 	}
 
-	sdep, err := stack.SerializeDeployment(snap, snap.SecretsManager)
+	sdep, err := stack.SerializeDeployment(snap, snap.SecretsManager, false /** showSecrets**/)
```
Comment (pgavlin) [inline cmd/state.go]:
```suggestion
	sdep, err := stack.SerializeDeployment(snap, snap.SecretsManager, false /* showSecrets */)
```

Comment (pgavlin) [reply cmd/state.go]: (and throughout)

File: cmd/stack_output.go (hunk 6 above)
Comment (stack72) [reply cmd/stack_output.go]: fixed!

File: cmd/stack_import.go (hunk 5 above)
Comment (stack72) [reply cmd/stack_import.go]: fixed!

File: cmd/state.go (hunk 7 above)
Comment (stack72) [reply cmd/state.go]: fixed!

File: pkg/backend/httpstate/backend.go [hunk 8]
```diff
@@ -54,6 +54,7 @@ import (
 	"github.com/pulumi/pulumi/pkg/util/result"
 	"github.com/pulumi/pulumi/pkg/util/retry"
 	"github.com/pulumi/pulumi/pkg/workspace"
+	"github.com/skratchdot/open-golang/open"
```
[the remark this answers is not in the document]
Comment (stack72) [reply pkg/backend/httpstate/backend.go]: fixed!

File: cmd/stack_output.go (hunk 6 above)
Comment (stack72) [reply cmd/stack_output.go]: fixed!

File: pkg/backend/httpstate/backend.go [hunk 9]
```diff
@@ -1177,18 +1178,47 @@ func (b *cloudBackend) ExportDeploymentForVersion(
...
 
 // exportDeployment exports the checkpoint file for a stack, optionally getting a previous version.
 func (b *cloudBackend) exportDeployment(
-	ctx context.Context, stackRef backend.StackReference, version *int) (*apitype.UntypedDeployment, error) {
+	ctx context.Context, stk backend.Stack, version *int, showSecrets bool) (*apitype.UntypedDeployment, error) {
+
+	stackRef := stk.Ref()
+	if showSecrets {
+		// get the snapshot
+		snap, err := b.getSnapshot(ctx, stk)
```
[the remark this answers is not in the document]
Comment (stack72) [reply pkg/backend/httpstate/backend.go]: Happy to update it - this was using the same codepath as the file backend

File: pkg/backend/httpstate/backend.go (hunk 1 above)
Comment (stack72) [reply pkg/backend/httpstate/backend.go]: I have not investigated this further up the stack - I was trying to be as close to the code as possible - we can of course try and move it further up but it could be more disruptive

File: pkg/resource/stack/deployment.go (hunk 2 above)
Comment (stack72) [reply pkg/resource/stack/deployment.go]: The normal stack import expects the text to be still in it's encrypted format - this change allows a stack that has been expected with secrets to be re-imported

File: pkg/backend/httpstate/backend.go (hunk 3 above)
Comment (stack72) [reply pkg/backend/httpstate/backend.go]: As above, we can move this further up but it was done to try and make this as less disruptive as possible

File: pkg/backend/httpstate/state.go (hunk 4 above)
Comment (stack72) [reply pkg/backend/httpstate/state.go]: The file codepath uses Stack not StackReference so I was solidifying on that path - we can of course go to StackReferences. I used stk here as it's used within the filestate codepath

File: pkg/backend/httpstate/backend.go (hunk 3 above)
Comment (pgavlin) [reply pkg/backend/httpstate/backend.go]: I'd prefer that we do this up the stack as well.

**Comment by NinoFloris** [comment]
We're running into to this, the workaround to migrate a stack - we're going from default secret provider to kms- is very tedious without this. Any update?

**Comment by stack72** [comment]
Hi @NinoFloris 

We paused this to concentrate on the 2.0 work so I will get back to this shortly

Paul

**Comment by stack72** [comment]
This allows to export and import secrets 

```
~/code/secrets
▶ pulumi stack export --show-secrets > test.json

~/code/secrets
▶ pulumi stack import --file test.json
Import successful.

~/code/secrets
▶ pulumi stack export > test.json

~/code/secrets
▶ pulumi stack import --file test.json
Import successful.
```

I can see the secrets in the output as well:

```
▶ cat test.json
{
    "version": 3,
    "deployment": {
        "manifest": {
            "time": "2020-05-05T22:31:33.276569+01:00",
            "magic": "7c37e5b973ae9a0866c7b0097ad75696c2881d030dcc17834b2ef914fe3ecba0",
            "version": "v2.2.0-alpha.1588707575+g7b446d6c.dirty"
        },
        "secrets_providers": {
            "type": "service",
            "state": {
                "url": "https://api.pulumi.com",
                "owner": "stack72",
                "project": "secrets",
                "stack": "dev"
            }
        },
        "resources": [
            {
                "urn": "urn:pulumi:dev::secrets::pulumi:pulumi:Stack::secrets-dev",
                "custom": false,
                "type": "pulumi:pulumi:Stack",
                "outputs": {
                    "secret": {
                        "4dabf18193072939515e22adb298388d": "1b47061264138c4ac30d75fd1eb44270",
                        "ciphertext": "AAABADjwqpv+jPnI8iB7ctNgF9tLe6cjry9KKIg8YNblA1JFiVIPzfcrmCsVhfXJ7hq+o73m9A=="
                    }
                }
            },
            {
                "urn": "urn:pulumi:dev::secrets::pulumi:providers:random::default_2_1_0",
                "custom": true,
                "id": "e4810fc7-d485-4cda-b62c-95e767296cdd",
                "type": "pulumi:providers:random",
                "inputs": {
                    "version": "2.1.0"
                },
                "outputs": {
                    "version": "2.1.0"
                }
            },
            {
                "urn": "urn:pulumi:dev::secrets::random:index/randomString:RandomString::my-string",
                "custom": true,
                "id": "O[tUh1E-TF\u003eW+P=t",
                "type": "random:index/randomString:RandomString",
                "inputs": {
                    "__defaults": [
                        "lower",
                        "minLower",
                        "minNumeric",
                        "minSpecial",
                        "minUpper",
                        "number",
                        "upper"
                    ],
                    "length": 16,
                    "lower": true,
                    "minLower": 0,
                    "minNumeric": 0,
                    "minSpecial": 0,
                    "minUpper": 0,
                    "number": true,
                    "special": true,
                    "upper": true
                },
                "outputs": {
                    "__meta": "{\"schema_version\":\"1\"}",
                    "id": "O[tUh1E-TF\u003eW+P=t",
                    "length": 16,
                    "lower": true,
                    "minLower": 0,
                    "minNumeric": 0,
                    "minSpecial": 0,
                    "minUpper": 0,
                    "number": true,
                    "result": {
                        "4dabf18193072939515e22adb298388d": "1b47061264138c4ac30d75fd1eb44270",
                        "ciphertext": "AAABADDSJYg7UC9JUNdQ3HdO+AXjxd2IlLN49Wd7qfXwea1T73F/DjsgndGoz5O6876LR0uN3Q=="
                    },
                    "special": true,
                    "upper": true
                },
                "parent": "urn:pulumi:dev::secrets::pulumi:pulumi:Stack::secrets-dev",
                "provider": "urn:pulumi:dev::secrets::pulumi:providers:random::default_2_1_0::e4810fc7-d485-4cda-b62c-95e767296cdd",
                "propertyDependencies": {
                    "length": null,
                    "special": null
                },
                "additionalSecretOutputs": [
                    "result"
                ]
            }
        ]
    }
}
```

File: pkg/resource/stack/deployment.go (hunk 2 above)
Comment (pgavlin) [reply pkg/resource/stack/deployment.go:520]: Can you explain this in a bit more detail? I'm also surprised that this is necessary.

Comment (stack72) [reply pkg/resource/stack/deployment.go:520]:
```
// If the decrypter is a cachingCrypter, insert the plain- and ciphertext into the cache with the
// new *resource.Secret as the key.
```

basically, without this work, we cannot reconstruct the secret
the existing codepath takes the secret, decrypts it and ensures it's correct
then reconstructs the secret
as the import (with secrets) doesn't have the secret, we need to re-encrypt the secret to be able to recreate the secret struct

File: pkg/cmd/pulumi/stack_export.go [hunk 10]
```diff
@@ -102,5 +137,7 @@ func newStackExportCmd() *cobra.Command {
 		&file, "file", "", "", "A filename to write stack output to")
 	cmd.PersistentFlags().StringVarP(
 		&version, "version", "", "", "Previous stack version to export. (If unset, will export the latest.)")
+	cmd.Flags().BoolVarP(
+		&showSecrets, "show-secrets", "", false, "Show secrets in export. Defaults to `false`")
```
[the remark this answers is not in the document]
Comment (stack72) [reply pkg/cmd/pulumi/stack_export.go:141]: Fixed!

File: pkg/backend/filestate/state.go [hunk 11]
```diff
@@ -172,7 +172,7 @@ func (b *localBackend) saveStack(name tokens.QName, snap *deploy.Snapshot, sm se
 	if filepath.Ext(file) == "" {
 		file = file + ext
 	}
-	chk, err := stack.SerializeCheckpoint(name, snap, sm)
+	chk, err := stack.SerializeCheckpoint(name, snap, sm /* showSecrets */, false)
```
[the remark this answers is not in the document]
Comment (stack72) [reply pkg/backend/filestate/state.go:175]: Good catch - you are correct that this does look weird!

File: pkg/resource/stack/deployment.go [hunk 12]
```diff
@@ -484,19 +505,34 @@ func DeserializePropertyValue(v interface{}, dec config.Decrypter) (resource.Pro
...
 				case resource.SecretSig:
-					ciphertext, ok := objmap["ciphertext"].(string)
-					if !ok {
-						return resource.PropertyValue{}, errors.New("malformed secret value: missing ciphertext")
+					ciphertext, ciperOk := objmap["ciphertext"].(string)
+					plaintext, plainOk := objmap["plaintext"].(string)
+					if !ciperOk && !plainOk {
+						return resource.PropertyValue{}, errors.New(
+							"malformed secret value: missing ciphertext or plaintext")
 					}
```
[the remark this answers is not in the document]
Comment (stack72) [reply pkg/resource/stack/deployment.go:513]: I would agree - that would indeed be an incorrect state - fixed!

File: pkg/resource/stack/deployment.go [hunk 13]
```diff
@@ -484,19 +505,34 @@ func DeserializePropertyValue(v interface{}, dec config.Decrypter) (resource.Pro
 					contract.Assert(isarchive)
 					return resource.NewArchiveProperty(archive), nil
 				case resource.SecretSig:
-					ciphertext, ok := objmap["ciphertext"].(string)
-					if !ok {
-						return resource.PropertyValue{}, errors.New("malformed secret value: missing ciphertext")
+					ciphertext, ciperOk := objmap["ciphertext"].(string)
```
Comment (pgavlin) [inline pkg/resource/stack/deployment.go:508]:
```suggestion
					ciphertext, cipherOk := objmap["ciphertext"].(string)
```

File: pkg/resource/stack/deployment.go [hunk 14]
```diff
@@ -484,19 +505,34 @@ func DeserializePropertyValue(v interface{}, dec config.Decrypter) (resource.Pro
...
+					plaintext, plainOk := objmap["plaintext"].(string)
+					if (!ciperOk && !plainOk) || (plainOk && ciperOk) {
```
Comment (pgavlin) [inline pkg/resource/stack/deployment.go:510]:
```suggestion
					if (!cipherOk && !plainOk) || (plainOk && cipherOk) {
```

File: pkg/resource/stack/deployment.go (hunk 13 above)
Comment (stack72) [reply pkg/resource/stack/deployment.go:508]: d'oh!

File: pkg/resource/stack/deployment.go (hunk 14 above)
Comment (stack72) [reply pkg/resource/stack/deployment.go:510]: d'oh!
